x/liquidity/amm/pool.go

```diff
@@ -174,99 +413,240 @@ func (pool *BasicPool) ProvidableXAmountOver(price sdk.Dec) (amt sdk.Int) {
+	sqrtPrice := utils.DecApproxSqrt(price)
+	// dy = ry - (sqrt((x + transX) * (y + transY) / P) - b)
+	amt = pool.ry.ToDec().Sub(sqrtXComp.Mul(sqrtYComp).QuoRoundUp(sqrtPrice).Sub(pool.transY)).TruncateInt()
+	if amt.GT(pool.ry) {
+		amt = pool.ry
+	}
+	if !amt.IsPositive() {
+		return zeroInt
+	}
+	return
 }
 
-// InitialPoolCoinSupply returns ideal initial pool coin minting amount.
-func InitialPoolCoinSupply(x, y sdk.Int) sdk.Int {
-	cx := len(x.BigInt().Text(10)) - 1 // characteristic of x
-	cy := len(y.BigInt().Text(10)) - 1 // characteristic of y
-	c := ((cx + 1) + (cy + 1) + 1) / 2 // ceil(((cx + 1) + (cy + 1)) / 2)
-	res := big.NewInt(10)
-	res.Exp(res, big.NewInt(int64(c)), nil) // 10^c
-	return sdk.NewIntFromBigInt(res)
+func (pool *RangedPool) Clone() Pool {
+	return &RangedPool{
+		rx:       pool.rx,
+		ry:       pool.ry,
+		ps:       pool.ps,
+		transX:   pool.transX,
+		transY:   pool.transY,
+		xComp:    pool.xComp,
+		yComp:    pool.yComp,
+		minPrice: pool.minPrice,
+		maxPrice: pool.maxPrice,
+	}
+}
+
+// Deposit returns accepted x and y coin amount and minted pool coin amount
+// when someone deposits x and y coins.
+func Deposit(rx, ry, ps, x, y sdk.Int) (ax, ay, pc sdk.Int) {
+	// Calculate accepted amount and minting amount.
+	// Note that we take as many coins as possible(by ceiling numbers)
+	// from depositor and mint as little coins as possible.
+
+	utils.SafeMath(func() {
+		rx, ry := rx.ToDec(), ry.ToDec()
+		ps := ps.ToDec()
+
+		// pc = floor(ps * min(x / rx, y / ry))
+		var ratio sdk.Dec
+		switch {
+		case rx.IsZero():
+			ratio = y.ToDec().QuoTruncate(ry)
+		case ry.IsZero():
+			ratio = x.ToDec().QuoTruncate(rx)
+		default:
+			ratio = sdk.MinDec(
+				x.ToDec().QuoTruncate(rx),
+				y.ToDec().QuoTruncate(ry),
+			)
+		}
+		pc = ps.MulTruncate(ratio).TruncateInt()
+
+		mintProportion := pc.ToDec().Quo(ps)             // pc / ps
+		ax = rx.Mul(mintProportion).Ceil().TruncateInt() // ceil(rx * mintProportion)
+		ay = ry.Mul(mintProportion).Ceil().TruncateInt() // ceil(ry * mintProportion)
+	}, func() {
+		ax, ay, pc = sdk.ZeroInt(), sdk.ZeroInt(), sdk.ZeroInt()
+	})
+
+	return
 }
 
-// MockPoolOrderSource demonstrates how to implement a pool OrderSource.
-type MockPoolOrderSource struct {
-	Pool
-	baseCoinDenom, quoteCoinDenom string
+// Withdraw returns withdrawn x and y coin amount when someone withdraws
+// pc pool coin.
+// Withdraw also takes care of the fee rate.
+func Withdraw(rx, ry, ps, pc sdk.Int, feeRate sdk.Dec) (x, y sdk.Int) {
+	if pc.Equal(ps) {
+		// Redeeming the last pool coin - give all remaining rx and ry.
+		x = rx
+		y = ry
+		return
+	}
+
+	utils.SafeMath(func() {
+		proportion := pc.ToDec().QuoTruncate(ps.ToDec())                             // pc / ps
+		multiplier := sdk.OneDec().Sub(feeRate)                                      // 1 - feeRate
+		x = rx.ToDec().MulTruncate(proportion).MulTruncate(multiplier).TruncateInt() // floor(rx * proportion * multiplier)
```

Comment:
How about the next change in order to reflect the math equation here?
If two `MulTruncate` are used, the the user will have more loss than the exact calculation.
```
x = rx.ToDec().MulTruncate(proportion.Mul(multiplier)).TruncateInt()
```
If this is already reviewed and passed, then it is okay for now.